File: java/examples/GoodServiceTest.java
package com.example.orders.application;

import com.example.orders.domain.Order;
import com.example.orders.domain.OrderStatus;
import com.example.orders.domain.exceptions.OrderNotFoundException;
import com.example.orders.domain.exceptions.OrderNotPayableException;
import com.example.orders.domain.exceptions.PaymentDeclinedException;
import com.example.orders.domain.exceptions.PaymentServiceException;
import com.example.orders.domain.ports.OrderRepository;
import com.example.orders.domain.ports.PaymentGateway;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.junit.jupiter.params.provider.ValueSource;

import java.math.BigDecimal;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

/**
 * Unit tests for {@link OrderPaymentService}.
 *
 * <p>Uses Mockito for gateway + repository fakes, AssertJ for fluent assertions.
 * Each test follows the Arrange → Act → Assert pattern.
 */
class OrderPaymentServiceTest {

  // ✅ Mocks declared at class level, re-initialised per test
  private OrderRepository orderRepository;
  private PaymentGateway paymentGateway;
  private OrderPaymentService service;

  @BeforeEach
  void setUp() {
    orderRepository = mock(OrderRepository.class);
    paymentGateway = mock(PaymentGateway.class);
    service = new OrderPaymentService(orderRepository, paymentGateway);
  }

  // ---------------------------------------------------------------------------
  // Helper builders
  // ---------------------------------------------------------------------------

  private Order buildPendingOrder(String orderId) {
    return Order.builder()
        .id(orderId)
        .userId("user-456")
        .total(new BigDecimal("100.00"))
        .currency("USD")
        .status(OrderStatus.PENDING)
        .build();
  }

  // ---------------------------------------------------------------------------
  // Happy path
  // ---------------------------------------------------------------------------

  @Nested
  @DisplayName("Given a valid pending order and successful payment")
  class HappyPath {

    @Test
    @DisplayName("should return payment ID and mark order as PAID")
    void processPayment_validCard_returnsPaymentId() {
      // Arrange
      Order order = buildPendingOrder("order-123");
      when(orderRepository.findById("order-123")).thenReturn(Optional.of(order));
      when(paymentGateway.charge(any(), anyString(), anyString())).thenReturn("pay-789");

      // Act
      String paymentId = service.processPayment("order-123", "tok_visa");

      // Assert
      assertThat(paymentId).isEqualTo("pay-789");
      verify(orderRepository).save(order);
      assertThat(order.getStatus()).isEqualTo(OrderStatus.PAID);
    }
  }

  // ---------------------------------------------------------------------------
  // Input validation
  // ---------------------------------------------------------------------------

  @Nested
  @DisplayName("Input validation")
  class Validation {

    @ParameterizedTest(name = "orderId = [{0}]")
    @NullAndEmptySource
    @ValueSource(strings = {"  "})
    @DisplayName("should reject blank or null orderId")
    void processPayment_blankOrderId_throwsIllegalArgument(String orderId) {
      assertThatThrownBy(() -> service.processPayment(orderId, "tok_visa"))
          .isInstanceOf(IllegalArgumentException.class);
      verifyNoInteractions(orderRepository, paymentGateway);
    }

    @ParameterizedTest(name = "cardToken = [{0}]")
    @NullAndEmptySource
    @ValueSource(strings = {"  "})
    @DisplayName("should reject blank or null cardToken")
    void processPayment_blankCardToken_throwsIllegalArgument(String cardToken) {
      assertThatThrownBy(() -> service.processPayment("order-123", cardToken))
          .isInstanceOf(IllegalArgumentException.class);
      verifyNoInteractions(orderRepository, paymentGateway);
    }
  }

  // ---------------------------------------------------------------------------
  // Order state errors
  // ---------------------------------------------------------------------------

  @Nested
  @DisplayName("Order state errors")
  class OrderStateErrors {

    @Test
    @DisplayName("should throw OrderNotFoundException when order does not exist")
    void processPayment_orderNotFound_throwsOrderNotFoundException() {
      when(orderRepository.findById(anyString())).thenReturn(Optional.empty());

      assertThatThrownBy(() -> service.processPayment("missing-id", "tok_visa"))
          .isInstanceOf(OrderNotFoundException.class);
      verifyNoInteractions(paymentGateway);
    }

    @Test
    @DisplayName("should throw OrderNotPayableException when order is already PAID")
    void processPayment_alreadyPaid_throwsOrderNotPayableException() {
      Order paidOrder = buildPendingOrder("order-123");
      paidOrder.setStatus(OrderStatus.PAID);
      when(orderRepository.findById("order-123")).thenReturn(Optional.of(paidOrder));

      assertThatThrownBy(() -> service.processPayment("order-123", "tok_visa"))
          .isInstanceOf(OrderNotPayableException.class);
      verifyNoInteractions(paymentGateway);
    }
  }

  // ---------------------------------------------------------------------------
  // Gateway failures
  // ---------------------------------------------------------------------------

  @Nested
  @DisplayName("Gateway failures")
  class GatewayFailures {

    @Test
    @DisplayName("should propagate PaymentDeclinedException and NOT save order")
    void processPayment_cardDeclined_throwsDeclinedException() {
      Order order = buildPendingOrder("order-123");
      when(orderRepository.findById("order-123")).thenReturn(Optional.of(order));
      when(paymentGateway.charge(any(), anyString(), anyString()))
          .thenThrow(new PaymentDeclinedException("insufficient_funds"));

      assertThatThrownBy(() -> service.processPayment("order-123", "tok_declined"))
          .isInstanceOf(PaymentDeclinedException.class);

      // ✅ Data integrity: order must not be saved in failed state
      verify(orderRepository, never()).save(any());
      assertThat(order.getStatus()).isEqualTo(OrderStatus.PENDING);
    }

    @Test
    @DisplayName("should wrap unexpected gateway errors in PaymentServiceException")
    void processPayment_gatewayUnavailable_throwsPaymentServiceException() {
      Order order = buildPendingOrder("order-123");
      when(orderRepository.findById("order-123")).thenReturn(Optional.of(order));
      when(paymentGateway.charge(any(), anyString(), anyString()))
          .thenThrow(new RuntimeException("connection refused"));

      assertThatThrownBy(() -> service.processPayment("order-123", "tok_visa"))
          .isInstanceOf(PaymentServiceException.class)
          .hasCauseInstanceOf(RuntimeException.class);
    }
  }
}

File: kotlin/examples/GoodServiceTest.kt
package com.example.orders.application

import com.example.orders.domain.Order
import com.example.orders.domain.OrderStatus
import io.mockk.*
import io.mockk.impl.annotations.MockK
import kotlinx.coroutines.test.runTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.*
import java.math.BigDecimal
import java.time.Instant

/**
 * Unit tests for [OrderPaymentService].
 *
 * Uses MockK for suspending function mocks and kotlinx-coroutines-test for
 * structured coroutine testing (runTest).
 *
 * Each test follows Arrange → Act → Assert.
 */
@DisplayName("OrderPaymentService")
class OrderPaymentServiceTest {

    @MockK lateinit var orderRepository: OrderRepository
    @MockK lateinit var paymentGateway: PaymentGateway

    private lateinit var service: OrderPaymentService

    @BeforeEach
    fun setUp() {
        MockKAnnotations.init(this)
        service = OrderPaymentService(orderRepository, paymentGateway)
    }

    // -------------------------------------------------------------------------
    // Test helpers
    // -------------------------------------------------------------------------

    private fun buildPendingOrder(orderId: String = "order-123") = Order(
        id = orderId,
        userId = "user-456",
        total = Money(BigDecimal("100.00"), "USD"),
        status = OrderStatus.PENDING,
        createdAt = Instant.now(),
    )

    // -------------------------------------------------------------------------
    // Happy path
    // -------------------------------------------------------------------------

    @Nested
    @DisplayName("Given a valid pending order and a successful payment gateway")
    inner class HappyPath {

        @Test
        @DisplayName("processPayment returns Success with paymentId")
        fun processPayment_validCard_returnsSuccess() = runTest {
            // Arrange
            val order = buildPendingOrder()
            coEvery { orderRepository.findById("order-123") } returns order
            coEvery { paymentGateway.charge(any(), any(), any()) } returns "pay-789"
            coJustRun { orderRepository.save(any()) }

            // Act
            val result = service.processPayment("order-123", "tok_visa")

            // Assert
            assertThat(result).isInstanceOf(PaymentResult.Success::class.java)
            assertThat((result as PaymentResult.Success).paymentId).isEqualTo("pay-789")
            coVerify { orderRepository.save(match { it.status == OrderStatus.PAID }) }
        }
    }

    // -------------------------------------------------------------------------
    // Precondition validation
    // -------------------------------------------------------------------------

    @Nested
    @DisplayName("Input validation")
    inner class Validation {

        @Test
        @DisplayName("blank orderId throws IllegalArgumentException")
        fun processPayment_blankOrderId_throwsIllegalArgument() = runTest {
            assertThrows<IllegalArgumentException> {
                service.processPayment("  ", "tok_visa")
            }
            coVerify(exactly = 0) { orderRepository.findById(any()) }
        }

        @Test
        @DisplayName("blank cardToken throws IllegalArgumentException")
        fun processPayment_blankCardToken_throwsIllegalArgument() = runTest {
            assertThrows<IllegalArgumentException> {
                service.processPayment("order-123", "")
            }
            coVerify(exactly = 0) { orderRepository.findById(any()) }
        }
    }

    // -------------------------------------------------------------------------
    // Order state errors
    // -------------------------------------------------------------------------

    @Nested
    @DisplayName("Order state errors")
    inner class OrderStateErrors {

        @Test
        @DisplayName("not found returns ServiceError")
        fun processPayment_orderNotFound_returnsServiceError() = runTest {
            coEvery { orderRepository.findById("missing") } returns null

            val result = service.processPayment("missing", "tok_visa")

            assertThat(result).isInstanceOf(PaymentResult.ServiceError::class.java)
            coVerify(exactly = 0) { paymentGateway.charge(any(), any(), any()) }
        }

        @Test
        @DisplayName("already-paid order returns ServiceError without charging")
        fun processPayment_alreadyPaid_returnsServiceError() = runTest {
            val order = buildPendingOrder().copy(status = OrderStatus.PAID)
            coEvery { orderRepository.findById(order.id) } returns order

            val result = service.processPayment(order.id, "tok_visa")

            assertThat(result).isInstanceOf(PaymentResult.ServiceError::class.java)
            coVerify(exactly = 0) { paymentGateway.charge(any(), any(), any()) }
        }
    }

    // -------------------------------------------------------------------------
    // Gateway failures — verify data integrity (no save on failure)
    // -------------------------------------------------------------------------

    @Nested
    @DisplayName("Gateway failures")
    inner class GatewayFailures {

        @Test
        @DisplayName("declined payment returns Declined without saving order")
        fun processPayment_cardDeclined_returnsDeclined() = runTest {
            val order = buildPendingOrder()
            coEvery { orderRepository.findById(order.id) } returns order
            coEvery { paymentGateway.charge(any(), any(), any()) } throws
                PaymentDeclinedException("insufficient_funds", "Card declined")

            val result = service.processPayment(order.id, "tok_declined")

            assertThat(result).isInstanceOf(PaymentResult.Declined::class.java)
            // ✅ Data integrity: order must NOT be saved after failure
            coVerify(exactly = 0) { orderRepository.save(any()) }
        }

        @Test
        @DisplayName("gateway exception returns ServiceError")
        fun processPayment_gatewayUnavailable_returnsServiceError() = runTest {
            val order = buildPendingOrder()
            coEvery { orderRepository.findById(order.id) } returns order
            coEvery { paymentGateway.charge(any(), any(), any()) } throws
                RuntimeException("connection refused")

            val result = service.processPayment(order.id, "tok_visa")

            assertThat(result).isInstanceOf(PaymentResult.ServiceError::class.java)
            coVerify(exactly = 0) { orderRepository.save(any()) }
        }
    }
}
